libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs

```diff
@@ -317,107 +310,171 @@ private static JToken EntityResultToInstanceJObject(string text, Result result)
+                        // get/create $instance
+                        JToken instanceRoot;
+                        if (!recognizerResult.Entities.TryGetValue("$instance", StringComparison.OrdinalIgnoreCase, out instanceRoot))
+                        {
+                            instanceRoot = new JObject();
+                            recognizerResult.Entities["$instance"] = instanceRoot;
+                        }
+
+                        // add instanceData
+                        JToken instanceData;
+                        if (!((JObject)instanceRoot).TryGetValue(result.Label.Name, StringComparison.OrdinalIgnoreCase, out instanceData))
+                        {
+                            instanceData = new JArray();
+                            instanceRoot[result.Label.Name] = instanceData;
+                        }
+
+                        ((JArray)instanceData).Add(EntityResultToInstanceJObject(text, result));
+                    }
                 }
             }
+            catch (ApplicationException)
+            {
+                return; // ---- This is a "Try" function, i.e., best effort only, no exception.
+            }
         }
 
-        private void InitializeModel()
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        private void InitializeModel(string modelFolder, string snapshotFile, ILabelResolver resolverMockup = null)
         {
-            if (_modelFolder == null)
+            if (resolverMockup != null)
             {
-#pragma warning disable CA2208 // Instantiate argument exceptions correctly
-                throw new ArgumentNullException("ModelFolder");
-#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+                _resolver = resolverMockup;
+                _isResolverMockup = true;
+                return;
             }
 
-            if (_snapshotFile == null)
             {
-#pragma warning disable CA2208 // Instantiate argument exceptions correctly
-                throw new ArgumentNullException("SnapshotFile");
-#pragma warning restore CA2208 // Instantiate argument exceptions correctly
-            }
+                if (string.IsNullOrWhiteSpace(modelFolder))
+                {
+                    throw new ArgumentNullException(nameof(modelFolder));
+                }
 
-            if (_resolver != null)
-            {
-                return;
+                if (string.IsNullOrWhiteSpace(snapshotFile))
+                {
+                    throw new ArgumentNullException(nameof(snapshotFile));
+                }
             }
 
-            var fullModelFolder = Path.GetFullPath(PathUtils.NormalizePath(_modelFolder));
+            var fullModelFolder = Path.GetFullPath(PathUtils.NormalizePath(modelFolder));
 
-            var orchestrator = orchestratorMap.GetOrAdd(fullModelFolder, path =>
+            _orchestrator = orchestratorMap.GetOrAdd(fullModelFolder, path =>
```

Comment:
What's the latency of this method on moderate sized models?

Reply:
The latency for that is just seconds. Loading the models are quick even though they can be up to 1GB. The actual time consuming part is building the resolver consuming some big snapshot file. Perhaps that's the reason the InitializeModel() function is called once inside RecognizeAsync(), which is async.

Reply:
what is the latency of the entire initializeModel?

Reply:
Asking because 1) this will get run on the first turn and not when the app service starts. This means that the first turn of the first conversation will be really slow, i.e. as long as this method takes.

Reply:
Ideally we'd want to transfer this time to application load somehow in the future. We don't need to resolve this now, but it is importnat. I'll create an issue for this.